rocAL/source/readers/file_source_reader.cpp

```diff
@@ -222,58 +273,51 @@ Reader::Status FileSourceReader::generate_file_names() {
     }
 
     if (_file_names.empty())
-        ERR("FileReader ShardID [" + TOSTR(_shard_id) + "] Did not load any file from " + _folder_path)
-
-    // the following code is required to make every shard the same size - required for the multi-gpu training
-    uint images_to_pad_shard = (ceil(_file_count_all_shards / _shard_count) * _shard_count) - _file_count_all_shards;
-    if (!images_to_pad_shard) {
-        for (uint i = 0; i < images_to_pad_shard; i++) {
-            if (get_file_shard_id() != _shard_id) {
-                _file_count_all_shards++;
-                incremenet_file_id();
-                continue;
+        WRN("FileReader ShardID [" + TOSTR(_shard_id) + "] Did not load any file from " + _folder_path)
+
+    auto dataset_size = _file_count_all_shards;
+    // Pad the _file_names with last element of the shard in the vector when _pad_last_batch_repeated is True
+    if (_shard_size > 0)
+        _padded_samples = _shard_size % _batch_count;
+    else
+        _padded_samples = largest_shard_size_without_padding() % _batch_count;
+    _last_batch_padded_size = (_batch_count > 1) ? (_batch_count - _padded_samples) : 0;
+
+    if (_pad_last_batch_repeated ==
+        true) {  // pad the last sample when the dataset_size is not divisible by
+                 // the number of shard's (or) when the shard's size is not
+                 // divisible by the batch size making each shard having equal
+                 // number of samples
+        uint total_padded_samples = 0; // initialize the total_padded_samples to 0
+        for (uint shard_id = 0; shard_id < _shard_count; shard_id++) {
+            uint start_idx = (dataset_size * shard_id) / _shard_count;
+            uint actual_shard_size_without_padding = std::floor((shard_id + 1) * dataset_size / _shard_count) - std::floor(shard_id * dataset_size / _shard_count);
+            uint largest_shard_size_without_padding = std::ceil(dataset_size * 1.0 / _shard_count);
+            auto start = _file_names.begin() + start_idx + total_padded_samples;
+            auto end = _file_names.begin() + start_idx + actual_shard_size_without_padding + total_padded_samples;
+            if (largest_shard_size_without_padding % _batch_count) {
+                _num_padded_samples = (largest_shard_size_without_padding - actual_shard_size_without_padding) + _batch_count - (largest_shard_size_without_padding % _batch_count);
+                _file_count_all_shards += _num_padded_samples;
+                total_padded_samples += _num_padded_samples;
+                for (uint num_padded_samples = 0; num_padded_samples < _num_padded_samples; num_padded_samples++) {
```

Comment:
instead of doing for loop, use C++ range function to insert range of values in one line